### ISSUE TITLE
Add IConsumer and IProducer support to StringTransport to facilitate testing producers and consumers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3609]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3609|
|Type|enhancement|
|Created|2009-01-08 22:33:47Z|
|Branch|https://github.com/twisted/twisted/tree/more-stringtransport-3609|

`StringTransport` is currently only an `ITransport` implementation.  Transports generally also implement `IConsumer` and `IPushProducer` and protocols often depend on this.  It would be good to be able to test such protocols with a fake transport implementation, and `StringTransport` is the most likely candidate for this.

<details><summary>Searchable metadata</summary>

```
trac-id__3609 3609
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_more_stringtransport_3609 branches/more-stringtransport-3609
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1231454027000000 1231454027000000
changetime__1231855130000000 1231855130000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25905]) Branching to 'more-stringtransport-3609'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Alright, there's an effort.  Most of this code was pulled out of the [#10610](https://github.com/twisted/twisted/issues/10610) branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

So the changes are cool, but it would be even better with some tests... What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25951]) some tests for the new functionality of StringTransport

refs [#3609](https://github.com/twisted/twisted/issues/3609)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Alright, I gave it a try.  I'm not sure if this is what you had in mind.  What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

That's perfect, thanks. Just one typo in `test_disconnectingTransportCannotResume`: `runtimeError`.

When fixed please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [25972]) Merge more-stringtransport-3609

Author: exarkun, itamar
Reviewer: therve
Fixes: [#3609](https://github.com/twisted/twisted/issues/3609)

Expand `twisted.test.proto_helpers.StringTransport` so that it implements `IConsumer`
and `IPushProducer` as well as `ITransport` (also declare all of these).  This makes
it a bit more like a real transport.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
